pkg/ipamd/datastore/data_store.go

```diff
@@ -692,42 +692,33 @@ func (ds *DataStore) GetEFAENIs() map[string]bool {
 // IsRequiredForWarmIPTarget determines if this ENI has warm IPs that are required to fulfill whatever WARM_IP_TARGET is
 // set to.
 func (ds *DataStore) isRequiredForWarmIPTarget(warmIPTarget int, eni *ENI) bool {
-	freeCidrs := 0
+	otherWarmIPs := 0
 	for _, other := range ds.eniPool {
-		if other.ID != eni.ID {
-			for _, otherPrefixes := range other.AvailableIPv4Cidrs {
-				if ((ds.isPDEnabled && otherPrefixes.IsPrefix == true) || (!ds.isPDEnabled && otherPrefixes.IsPrefix == false)) && otherPrefixes.AssignedIPv4AddressesInCidr() == 0 {
-					freeCidrs++
-				}
-			}
-		}
-	}
-	if ds.isPDEnabled {
-		_, numIPsPerPrefix, _ := GetPrefixDelegationDefaults()
-		warmIPTarget = DivCeil(warmIPTarget, numIPsPerPrefix)
-	}
-	return freeCidrs < warmIPTarget
+                if other.ID != eni.ID {
+                       for _, otherPrefixes := range other.AvailableIPv4Cidrs {
+                               if (ds.isPDEnabled && otherPrefixes.IsPrefix == true) || (!ds.isPDEnabled && otherPrefixes.IsPrefix == false) {
+                                       otherWarmIPs += otherPrefixes.Size() - otherPrefixes.AssignedIPv4AddressesInCidr() 
+                               }
+                       }
+               }
+       }
+       return otherWarmIPs < warmIPTarget
```

Comment:
Might be a problem in PD mode since we round up the `WARM_IP_TARGET` to the next multiple of 16 (with current /28 Prefix). Say, our WARM_IP_TARGET is `14` and we have `15` free IPs spread across other ENIs. `otherWarmIPs < warmIPTarget` will return false and we will delete that ENI. The reconciler will then see that we need one more prefix once it rounds up the WARM_IP_TARGET (`14`) to `16` and will create one more ENI and attach a prefix to it and this might happen in loop if system stays in this state for a while. We should adjust the `WARM_IP_TARGET` count in PD mode.

Reply:
Yeah true, for certain cases this will try free an ENI.